src/clients/crux-wallet-onboarding-client.ts

```diff
@@ -0,0 +1,151 @@
+const log = getLogger(__filename);
+// DDD Experimental Stuff
+export interface ICruxOnBoardingClientOptions {
+    cacheStorage?: StorageService;
+    configKey?: string|IKeyManager;
+    domain?: string;
+    assetMapping?: IClientAssetMapping;
+}
+export class CruxOnBoardingClient {
+    private _initPromise: Promise<void>;
+    private _cruxDomainRepository: ICruxDomainRepositoryConstructor;
+    private _cruxAssetTranslatorRepository: ICruxAssetTranslatorRepositoryConstructor;
+    private _configKeyManager?: IKeyManager;
+    private _cruxDomain?: CruxDomain;
+    private _cruxAssetTranslator?: CruxAssetTranslator;
+    private _domainContext?: string;
+    private _assetMapping?: IClientAssetMapping;
+    constructor(options: ICruxOnBoardingClientOptions) {
+        setCacheStorage(options.cacheStorage || new InMemStorage());
+        // set the repository constructors to be used further in the client
+        this._cruxDomainRepository = BlockstackCruxDomainRepository;
+        this._cruxAssetTranslatorRepository = BlockstackCruxAssetTranslatorRepository;
+        // set the keyManagers if available
+        this._configKeyManager =  typeof options.configKey === "string" ? new ManualKeyManager(options.configKey) : options.configKey;
+        // set the domainContext if domain is provided
+        this._domainContext = options.domain;
+        this._initPromise = this._init();
+        log.info("CruxOnBoardingClient initialised");
+    }
+    set domain(domain: string) {
+        this._domainContext = domain;
+    }
+    public isCruxDomainAvailable = async (domain: string): Promise<boolean> => {
+        return new this._cruxDomainRepository().find(domain);
+    }
+    public registerCruxDomain = async (domain: string): Promise<void> => {
+        // TODO: implementation of auto registration of domain on blockchain
+        throw ErrorHelper.getPackageError(null, PackageErrorCode.IsNotSupported);
+    }
+    public getCruxDomainState = async (): Promise<DomainRegistrationStatus> => {
+        await this._initPromise;
+        return this._getCruxDomain().status;
+    }
+    public getNameServiceConfig = async () => {
+        await this._initPromise;
+        return this._getCruxDomain().nameServiceConfig;
+    }
+    public getAssetMapping = async (): Promise<IClientAssetMapping> => {
+        await this._initPromise;
+        return this._getCruxAssetTranslator().assetMapping;
+    }
+    public getAssetList = async (): Promise<IGlobalAssetList> => {
+        await this._initPromise;
+        return this._getCruxAssetTranslator().assetList;
+    }
+    public putNameServiceConfig = async (newNameServiceConfig: IBlockstackServiceInputOptions): Promise<void> => {
+        await this._initPromise;
+        const cruxDomain = this._getCruxDomain();
+        cruxDomain.nameServiceConfig = newNameServiceConfig;
+        this._cruxDomain = await this._getCruxDomainRepository().save(cruxDomain, this._getConfigKeyManager());
```

Comment:
Why instantiate the repository while using its methods? can we move it to the constructor?

Reply:
the repository initialisation will be different based on the CruxDomain set in memory.